File: webapp/go/top_handler.go
package main

import (
	"database/sql"
	"errors"
	"net/http"
	"strings"

	"github.com/labstack/echo/v4"
)

type Tag struct {
	Id   int64  `json:"id"`
	Name string `json:"name"`
	// CreatedAt is the created timestamp that forms an UNIX time.
	CreatedAt int64 `json:"created_at"`
}

type TagModel struct {
	Id   int64  `db:"id"`
	Name string `db:"name"`
	// CreatedAt is the created timestamp that forms an UNIX time.
	CreatedAt int64 `db:"created_at"`
}

type TagsResponse struct {
	Tags []*Tag `json:"tags"`
}

func getTagHandler(c echo.Context) error {
	ctx := c.Request().Context()

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	var tagModels []*TagModel
	if err := tx.SelectContext(ctx, &tagModels, "SELECT * FROM tags"); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to commit")
	}

	tags := make([]*Tag, len(tagModels))
	for i := range tagModels {
		tags[i] = &Tag{
			Id:        tagModels[i].Id,
			Name:      tagModels[i].Name,
			CreatedAt: tagModels[i].CreatedAt,
		}
	}
	return c.JSON(http.StatusOK, &TagsResponse{
		Tags: tags,
	})
}

// 配信者のテーマ取得API
// GET /theme
func getStreamerThemeHandler(c echo.Context) error {
	ctx := c.Request().Context()

	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		c.Logger().Printf("verifyUserSession: %+v\n", err)
		return err
	}

	hostHeader := c.Request().Host

	username := strings.Split(hostHeader, ".")[0]

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	userModel := UserModel{}
	err = tx.GetContext(ctx, &userModel, "SELECT id FROM users WHERE name = ?", username)
	if errors.Is(err, sql.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, err.Error())
	}
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	themeModel := ThemeModel{}
	if err := tx.GetContext(ctx, &themeModel, "SELECT * FROM themes WHERE user_id = ?", userModel.Id); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to commit")
	}

	theme := Theme{
		Id:        themeModel.Id,
		DarkMode:  themeModel.DarkMode,
		CreatedAt: themeModel.CreatedAt,
	}

	return c.JSON(http.StatusOK, theme)
}

File: webapp/node/src/handlers/top-handler.ts
import { Context } from 'hono'
import { RowDataPacket } from 'mysql2/promise'
import { HonoEnvironment } from '../types/application'
import { verifyUserSessionMiddleware } from '../middlewares/verify-user-session-middleare'
import { TagsModel, ThemeModel, UserModel } from '../types/models'
import { throwErrorWith } from '../utils/throw-error-with'

// GET /api/tag
export const getTagHandler = async (
  c: Context<HonoEnvironment, '/api/tag'>,
) => {
  const conn = await c.get('pool').getConnection()
  await conn.beginTransaction()

  try {
    const [tags] = await conn
      .execute<(TagsModel & RowDataPacket)[]>('SELECT * FROM tags')
      .catch(throwErrorWith('failed to get tags'))

    await conn.commit().catch(throwErrorWith('failed to commit'))

    const tagResponses = []
    for (const tag of tags) {
      tagResponses.push({
        id: tag.id,
        name: tag.name,
      })
    }

    return c.json({ tags: tagResponses })
  } catch (error) {
    await conn.rollback()
    return c.text(`Internal Server Error\n${error}`, 500)
  } finally {
    await conn.rollback()
    conn.release()
  }
}

// 配信者のテーマ取得API
// GET /api/user/:username/theme
export const getStreamerThemeHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/user/:username/theme'>) => {
    const username = c.req.param('username')

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[user]] = await conn
        .execute<(Pick<UserModel, 'id'> & RowDataPacket)[]>(
          'SELECT id FROM users WHERE name = ?',
          [username],
        )
        .catch(throwErrorWith('failed to get user'))

      if (!user) {
        await conn.rollback()
        return c.text('not found user that has the given username', 404)
      }

      const [[theme]] = await conn
        .execute<(ThemeModel & RowDataPacket)[]>(
          'SELECT * FROM themes WHERE user_id = ?',
          [user.id],
        )
        .catch(throwErrorWith('failed to get user theme'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      const themeResponse = {
        id: theme.id,
        dark_mode: !!theme.dark_mode,
      }

      return c.json(themeResponse)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]
